nservicebus/messaging/messages-as-interfaces.md

```diff
@@ -1,21 +1,97 @@
 ---
 title: Messages as Interfaces
-summary: NServiceBus supports messages being defined as interfaces
+summary: Define messages as interfaces in NServiceBus to support dynamic dispatch and polymorphic routing scenarios
 component: core
-reviewed: 2024-01-23
+reviewed: 2024-03-28
 redirects:
 - nservicebus/messages-as-interfaces
 ---
 
-NServiceBus supports using interfaces as messages to allow multiple inheritance, where one message can inherit from varieties of other messages. This is useful for solving certain message evolution scenarios.
+NServiceBus supports using interfaces as messages to allow dynamic dispatch and polymorphic message routing scenarios. This makes systems more flexible and maintainable by encouraging messages and handlers to be decoupled from each other. It's useful for evolving systems to allow adding new features without affecting existing functionality. It's best described using an example.
 
-Imagine that the business logic represents a state machine with states X and Y. When the system gets into state X, it publishes the message `EnteredStateX`. When the system gets into state Y, it publishes the message `EnteredStateY`. (For more information on how to publish a message, see below.)
+## Using NServiceBus for dynamic dispatch
 
-In the next version of the system, a new state Z is added, which represents the co-existence of both X and Y states. This can be achieved by defining a message `EnteredStateZ` which inherits from `EnteredStateX` and `EnteredStateY`.
+Consider the following handler stubs for a `UserCreated` message which inherits from the `IEvent` stub interface:
 
-When the system publishes the `EnteredStateZ` event, the clients that are subscribed to either `EnteredStateX` or `EnteredStateY` are notified.
+```csharp
+public class SaveUser : IHandleMessages<UserCreated>
+{
+    public async Task Handle(UserCreated message, IMessageHandlerContext) {
+        // Do stuff when the user is created
+    }
+}
 
-Without the ability to inherit a message from other message types, composition techniques would be required, thereby preventing the infrastructure from automatically routing messages to pre-existing subscribers of the composed messages.
+public class Audit : IHandleMessages<IMessage>
+{
+    public async Task Handle(IMessage message, IMessageHandlerContext context) {
+        // Audit the message
+    }
+}
+```
+
+When a message of type `UserCreated` is published, both handlers will be invoked independently which in itself is useful. However, this idea becomes more powerful as the system evolves and more functionality is added.
+
+Assume there is a separate process that needs to happen if a user is created as part of a campaign. For example, say the subscriber must register statistics about the campaign. In NServiceBus, this can be done by defining a new event and handler:
```

Comment:
```suggestion
Meanwhile, imagine there is a separate process that needs to happen if a user is created as part of a campaign. For example, say the subscriber must register statistics about the campaign. In NServiceBus, this can be done by defining a new event and handler:
```

Does "campaign" mean something in .Net or something? I find it gets used a lot, but it means nothing to me.

Reply:
Not that I'm aware of, to me it means an ad campaign, so it gets used a lot when you talk about getting messages from users, because it is a real life example on how someone will use the system.

Reply:
Ok, so "ad campaign" does mean something to me.

```suggestion
Assume there is a separate process that needs to happen if a user is created as part of an advertising campaign. For example, say the subscriber must register statistics about the campaign. In NServiceBus, this can be done by defining a new event and handler:
```